network/framework/src/application/storage.rs

```diff
@@ -91,6 +103,69 @@ impl PeersAndMetadata {
         all_peers
     }
 
+    /// Return copy of list of peers if generation is different than previously held.
+    /// May optionally filter for connected peers only.
+    /// May optionally filter on match for any of a set of ProtocolId.
+    pub fn get_all_peers_generational(
+        &self,
+        generation: u32,
+        require_connected: bool,
+        protocol_ids: &[ProtocolId],
+    ) -> Option<(Vec<PeerNetworkId>, u32)> {
+        let generation_test = self.generation.load(Ordering::SeqCst);
+        if generation == generation_test {
+            return None;
+        }
+        let mut all_peers = Vec::new();
+        let read = self.peers_and_metadata.read();
+        for (network_id, peers) in read.iter() {
+            for (peer_id, peer_metadata) in peers.iter() {
+                if require_connected && !peer_metadata.is_connected() {
+                    continue;
+                }
+                if !protocol_ids.is_empty() && !peer_metadata.supports_any_protocol(protocol_ids) {
+                    continue;
+                }
+                let peer_network_id = PeerNetworkId::new(*network_id, *peer_id);
+                all_peers.push(peer_network_id);
+            }
+        }
+        let generation_actual = self.generation.load(Ordering::SeqCst);
+        Some((all_peers, generation_actual))
+    }
+
+    /// Return copy of list of peers if generation is different than previously held.
+    /// May optionally filter for connected peers only.
+    /// May optionally filter on match for any of a set of ProtocolId.
+    pub fn get_all_peers_and_metadata_generational(
```

Comment:
Maybe I'm missing the point, but it's tough to see the benefits of this change without actually seeing the call sites change and performance numbers to justify the change. 

Also, if in the next PR we're providing "notifications for connection events", why do we need generational support at all? i.e., applications won't be polling the peers and metadata struct to discover new peers, so reads of peers and metadata information will only be required for "special cases" (e.g., optimizations based on peer metadata), and these reads can already be serviced through the existing cache, no?

Reply:
The subscription service is just for connects and disconnects; updates to metadata don't generate an event.

Reply:
Hmm... that should be fine. Applications don't currently rely on this. Instead, they fetch and read the data when required. It's infrequent for mempool, and more frequent for state sync (but it depends on load). I'm not sure sending events for "metadata changes" provides any benefit.